File: backend.v2/backend.v2/Models/Book.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;
using backend.v2.Configuration.Attributes;

namespace backend.v2.Models
{
    /// <summary>
    /// Перечисление для определения статуса книги.
    /// Книга может находиться в статусах "К прочтению", "Читаю" или "Буду читать".
    /// </summary>
    public enum Status
    {
        ToRead = 0,
        InProgress = 1,
        Done = 2
    }
    
    /// <summary>
    /// Перечисление для определения типа книги.
    /// Книга может иметь тип "Бумажная", "Электронная", "Аудиокнига".
    /// </summary>
    public enum Type
    {
        Paper = 0,
        Electronic = 1,
        Audio = 2
    }

    public class Book: IEntity
    {
        /// <summary>
        /// Уникальный идентификатор книги.
        /// </summary>
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        [Key]
        [Column(TypeName = "uuid")]
        public Guid? Guid { get; set; }
        
        /// <summary>
        /// Название книги.
        /// </summary>
        [Column(TypeName = "varchar(512)")]
        [Required]
        public string Name { get; set; }
        
        /// <summary>
        /// Список авторов.
        /// </summary>
        [Column(TypeName = "varchar(512)[]")]
        public string[] Authors { get; set; }
        
        /// <summary>
        /// Статус книги.
        /// </summary>
        public Status? Status { get; set; }
        
        /// <summary>
        /// Список тегов книги.
        /// </summary>
        [Column(TypeName = "varchar(256)[]")]
        public string[] Tags { get; set; }
        
        /// <summary>
        /// Количество прочитаных страниц/прослушанных минут.
        /// </summary>
        public short? DoneUnits { get; set; }
        
        /// <summary>
        /// Количество страниц/минут в книге.
        /// </summary>
        public short? TotalUnits { get; set; }

        [JsonIgnore]
        [SwaggerIgnore]
        public Collection Collection {get; set;}
        
        /// <summary>
        /// Коллекция к которой принадлежит книга.
        /// </summary>
        public Guid? CollectionGuid {get; set;}

        /// <summary>
        /// Номер книги в коллекции.
        /// </summary>
        public short? CollectionOrder {get; set;}
        
        /// <summary>
        /// Название жанра.
        /// </summary>
        [Column(TypeName = "varchar(256)")]
        public string Genre { get; set; }
        
        /// <summary>
        /// Год даты начала чтения.
        /// </summary>
        public short? StartDateYear { get; set; }
        
        /// <summary>
        /// Месяц даты начала чтения.
        /// </summary>
        public short? StartDateMonth { get; set; }
        
        /// <summary>
        /// День даты начала чтения.
        /// </summary>
        public short? StartDateDay { get; set; }
        
        /// <summary>
        /// Год даты окончания чтения.
        /// </summary>
        public short? EndDateYear { get; set; }
        
        /// <summary>
        /// Месяц даты окончания чтения.
        /// </summary>
        public short? EndDateMonth { get; set; }
        
        /// <summary>
        /// День даты окончания чтения.
        /// </summary>
        public short? EndDateDay { get; set; }
        
        /// <summary>
        /// Год выпуска книги.
        /// </summary>
        public short? Year { get; set; }
        
        /// <summary>
        /// Время последнего изменения книги.
        /// </summary>
        [Column(TypeName = "timestamptz")]
        public DateTime? ModifyDate { get; set; }
        
        /// <summary>
        /// Время создания книги.
        /// </summary>
        [Column(TypeName = "timestamptz")]
        public DateTime? CreateDate { get; set; }
        
        /// <summary>
        /// Время удаления книги.
        /// </summary>
        [Column(TypeName = "timestamptz")]
        public DateTime? DeleteDate { get; set; }

        /// <summary>
        /// Способ расчета прогресса чтения.
        /// Может принимать значения "Done", "Left".
        /// Используется "Done" по умолчанию.
        /// </summary>
        [Column(TypeName = "varchar(16)")]
        public string ProgressType { get; set; }

        /// <summary>
        /// Дата начала прочтения.
        /// </summary>
        public DateTime? StartDate { 
            get {
                if (!this.StartDateYear.HasValue) return null;
                
                return new DateTime(this.StartDateYear ?? 1, this.StartDateMonth ?? 1, this.StartDateDay ?? 1);
            }
        }
        

        /// <summary>
        /// Дата окончания прочтения.
        /// </summary>
        public DateTime? EndDate { 
            get {
                if (!this.EndDateYear.HasValue) return null;
                
                return new DateTime(this.EndDateYear ?? 1, this.EndDateMonth ?? 1, this.EndDateDay ?? 1);
            }
        }
        
        /// <summary>
        /// Маркер того, что коллекция уделена.
        /// Всегда `true`, когда заполнена `DeleteDate` и, соответственно, всегда `false`, когда ее нет.
        /// </summary>
        public bool Deleted {
            get {
                return this.DeleteDate != null;
            }
        }
        
        
        /// <summary>
        /// Тип книги.
        /// </summary>
        public Type? Type { get; set; }
        
        /// <summary>
        /// Заметка о книге.
        /// </summary>
        public string Note { get; set; }
        
        
        /// <summary>
        /// `Id` пользователя, владеющего книгой.
        /// </summary>
        public long UserId { get; set; }
        
        [SwaggerIgnore]
        [JsonIgnore]
        public User User { get; set; }
    }
}

File: frontend.v3/src/app/modules/book/models/book.ts
import { Entity } from '../../../main/models/entity';
import { Collection } from '../../collection/models/collection';
import { BookData } from './book-data';
import { BookDate } from './book-date';
import { BookStatus } from './book-status';
import { BookType } from './book-type';
import { ProgressAlgorithmType } from './progress-algorithm-type';

export class Book extends Entity {
  public name: string;
  public authors: string[];
  public year?: number;
  public status: BookStatus;
  public tags: string[];
  public totalUnits: number;
  public doneUnits: number;
  public genre?: string;
  public collection?: Collection = null;
  public collectionGuid?: string;
  public collectionOrder?: number;
  public started: BookDate;
  public startDate?: Date;
  public finished: BookDate;
  public endDate?: Date;
  public type: BookType;
  public note?: string;
  public progressType: ProgressAlgorithmType;
  public rereadingBookGuid?: string;
  public rereadedBy: string[];

  public constructor(data: BookData) {
    super(data);

    this.name = data.name;
    this.authors = data.authors || [];
    this.year = data.year;
    this.status = data.status;
    this.tags = data.tags || [];
    this.totalUnits = data.totalUnits || 0;
    this.doneUnits = data.doneUnits || 0;
    this.genre = data.genre;
    this.collectionGuid = data.collectionGuid;
    this.collectionOrder = data.collectionOrder;
    this.started = {
      year: data.startDateYear,
      month: data.startDateMonth,
      day: data.startDateDay,
    };
    this.startDate = this.getDate(data.startDate, this.started);
    this.finished = {
      year: data.endDateYear,
      month: data.endDateMonth,
      day: data.endDateDay,
    };
    this.endDate = this.getDate(data.endDate, this.finished);
    this.type = data.type;
    this.note = data.note;
    this.progressType = data.progressType || ProgressAlgorithmType.Done;
    this.rereadingBookGuid = data.rereadingBookGuid || null;
    this.rereadedBy = data.rereadedBy || [];
  }

  public get done(): number {
    if (this.totalUnits == null || this.doneUnits == null) {
      return null;
    }
    if (this.progressType === ProgressAlgorithmType.Left) {
      return this.totalUnits - this.doneUnits;
    } else {
      return this.doneUnits;
    }
  }

  public set done(v: number) {
    if (this.progressType === ProgressAlgorithmType.Left) {
      this.doneUnits = this.totalUnits - v;
    } else {
      this.doneUnits = v;
    }
  }

  public get progressPercents(): number {
    if (this.totalUnits && this.doneUnits) {
      return Math.floor((this.doneUnits / this.totalUnits) * 100);
    } else {
      return 0;
    }
  }

  public toString(): string {
    return this.authors.concat([this.name]).join('|');
  }

  private getDate(date: string | Date, bDate: BookDate): Date {
    if (date) {
      return new Date(date);
    } else if (bDate.day || bDate.month || bDate.year) {
      const month = bDate.month ? bDate.month - 1 : 0;
      return new Date(bDate.year, month, bDate.day || 1);
    } else {
      return null;
    }
  }
}
